core/node/consensus/src/batch.rs

```
@@ -0,0 +1,266 @@
+        let this_batch = LastBlockWitness::load(ctx, number, pool, tree)
+            .await
+            .with_wrap(|| format!("LastBlockWitness::make({number})"))?;
+        let mut conn = pool.connection(ctx).await.wrap("connection()")?;
+        let this = Self {
+            blocks: conn
+                .payloads(
+                    ctx,
+                    std::ops::Range {
+                        start: prev_batch.last_block() + 1,
+                        end: this_batch.last_block() + 1,
+                    },
+                )
+                .await
+                .wrap("payloads()")?,
+            prev_batch,
+            this_batch,
+        };
+        Ok(this)
+    }
+
+    /// Verifies the L1Batch and witness against the commitment.
+    /// WARNING: the following fields of the payload are not currently verified:
+    /// * `l1_gas_price`
+    /// * `l2_fair_gas_price`
+    /// * `fair_pubdata_price`
+    /// * `virtual_blocks`
+    /// * `operator_address`
+    /// * `protocol_version` (present both in payload and witness, but neither has a commitment)
+    pub(crate) fn verify(&self, comm: &L1BatchCommit) -> anyhow::Result<()> {
+        let (last_number, last_hash) = self.this_batch.verify(&comm.this_batch)?;
+        let (mut prev_number, mut prev_hash) = self.prev_batch.verify(&comm.prev_batch)?;
+        anyhow::ensure!(
+            self.prev_batch
+                .info
+                .batch_number
+                .checked_add(1)
+                .context("batch_number overflow")?
+                == u64::from(comm.number.0)
+        );
+        anyhow::ensure!(self.this_batch.info.batch_number == u64::from(comm.number.0));
+        for (i, b) in self.blocks.iter().enumerate() {
+            anyhow::ensure!(b.l1_batch_number == comm.number);
+            anyhow::ensure!(b.protocol_version == self.this_batch.protocol_version);
+            anyhow::ensure!(b.last_in_batch == (i + 1 == self.blocks.len()));
```

do we also want to add `anyhow::ensure` for the intermediate results of hashes, etc? It does not give any additional security, but allows to fail fast

---

sg